compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmStringConcatenationLowering.kt

```diff
@@ -0,0 +1,146 @@
+
+internal val jvmStringConcatenationLowering = makeIrFilePhase(
+    ::JvmStringConcatenationLowering,
+    name = "StringConcatenation",
+    description = "Replace IrStringConcatenation with string builders"
+)
+
+/**
+ * This lowering pass replaces [IrStringConcatenation]s with StringBuilder appends.
+ */
+private class JvmStringConcatenationLowering(val context: JvmBackendContext) : FileLoweringPass, IrElementTransformerVoidWithContext() {
+    override fun lower(irFile: IrFile) = irFile.transformChildrenVoid()
+
+    private val typesWithSpecialAppendFunction = context.irBuiltIns.primitiveIrTypes + context.irBuiltIns.stringType
+
+    private val nameToString = Name.identifier("toString")
+    private val nameAppend = Name.identifier("append")
+
+    private val stringBuilder = context.ir.symbols.stringBuilder.owner
+
+    private val constructor = stringBuilder.constructors.single {
+        it.valueParameters.size == 0
+    }
+
+    private val toStringFunction = stringBuilder.functions.single {
+        it.valueParameters.size == 0 && it.name == nameToString
+    }
+
+    private val defaultAppendFunction = stringBuilder.functions.single {
+        it.name == nameAppend &&
+                it.valueParameters.size == 1 &&
+                it.valueParameters.single().type.isNullableAny()
+    }
+
+    private val appendFunctions: Map<IrType, IrSimpleFunction?> =
+        typesWithSpecialAppendFunction.map { type ->
+            type to stringBuilder.functions.toList().atMostOne {
+                it.name == nameAppend && it.valueParameters.singleOrNull()?.type == type
+            }
+        }.toMap()
+
+    private fun typeToAppendFunction(type: IrType): IrSimpleFunction {
+        return appendFunctions[type] ?: defaultAppendFunction
+    }
+
+    private fun JvmIrBuilder.callToString(expression: IrExpression): IrExpression {
+        if (expression.type.isString())
+            return expression
+
+        // There is no String.valueOf for bytes or shorts.
+        val argument = if (expression.type.isByte() || expression.type.isShort())
+            irImplicitCast(expression, context.irBuiltIns.intType)
+        else expression
+
+        val argumentType = if (argument.type.isPrimitiveType()) argument.type else context.irBuiltIns.anyNType
+
+        val function = backendContext.ir.symbols.javaLangString.functions.single {
+            it.owner.name.asString() == "valueOf" && it.owner.valueParameters.singleOrNull()?.type == argumentType
+        }
+
+        return irCall(function).apply {
+            putValueArgument(0, argument)
+        }
+    }
+
+    private fun JvmIrBuilder.lowerInlineClassArgument(expression: IrExpression): IrExpression {
+        if (expression.type.unboxInlineClass() == expression.type)
+            return expression
+
+        val toStringFunction = expression.type.classOrNull?.owner?.functions?.singleOrNull {
+            it.name.asString() == "toString" && it.dispatchReceiverParameter != null
```

Comment:
Maybe either use the `nameToString` value here as well, or inline it

Reply:
Also, there's similar code in `FlattenStringConcatenationLowering`, would be nice to extract to a utility in backend.common

Reply:
Done!